ffb-server/src/main/java/com/fumbbl/ffb/server/handler/ServerCommandHandlerJoinApproved.java

```diff
@@ -251,6 +251,8 @@ private void sendTeamList(GameState pGameState, InternalServerCommandJoinApprove
 					.add(new FumbblRequestLoadTeamList(pGameState, pJoinApprovedCommand.getCoach(), pSession));
 		} else {
 			TeamList teamList = new TeamList();
+			// In STANDALONE mode, we need to initialize the rules before we start parsing teams.
+			pGameState.getGame().initializeRules();
```

Comment:
Did you get an error that required this change? So far I had no issues with rules being initialized too late in standalone and I do not see any other changes in this PR that would alter that behaviour?

Reply:
The error happens when you are parsing the new XML format from the FUMBBL API. It doesn't happen on the older roster files, which is probably why you haven't seen it. This is the stacktrace:

```
2024-10-17T08:40:16.336 -------- ---- com.fumbbl.ffb.FantasyFootballException: Error deserializing roster for id 1158751
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.server.RosterCache.getRosterForTeam(RosterCache.java:65)
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.server.GameCache.updateRoster(GameCache.java:246)
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.server.GameCache.lambda$getTeamsForCoach$1(GameCache.java:304)
2024-10-17T08:40:16.336 -------- ---- 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
2024-10-17T08:40:16.336 -------- ---- 	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
2024-10-17T08:40:16.336 -------- ---- 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
2024-10-17T08:40:16.336 -------- ---- 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
2024-10-17T08:40:16.336 -------- ---- 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)
2024-10-17T08:40:16.336 -------- ---- 	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
2024-10-17T08:40:16.336 -------- ---- 	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.server.GameCache.getTeamsForCoach(GameCache.java:304)
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.server.handler.ServerCommandHandlerJoinApproved.sendTeamList(ServerCommandHandlerJoinApproved.java:254)
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.server.handler.ServerCommandHandlerJoinApproved.handleCommand(ServerCommandHandlerJoinApproved.java:110)
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.server.handler.ServerCommandHandlerFactory.handleCommand(ServerCommandHandlerFactory.java:46)
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.server.net.ServerCommunication.handleCommandInternal(ServerCommunication.java:131)
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.server.net.ServerCommunication.run(ServerCommunication.java:106)
2024-10-17T08:40:16.336 -------- ---- 	at java.base/java.lang.Thread.run(Thread.java:1583)
2024-10-17T08:40:16.336 -------- ---- Caused by: com.fumbbl.ffb.FantasyFootballException: Parsing error.
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.xml.XmlHandler.parse(XmlHandler.java:92)
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.server.RosterCache.getRosterForTeam(RosterCache.java:61)
2024-10-17T08:40:16.336 -------- ---- 	... 16 more
2024-10-17T08:40:16.336 -------- ---- Caused by: java.lang.NullPointerException: Cannot invoke "java.util.Map.get(Object)" because "this.factories" is null
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.model.GameRules.getFactory(GameRules.java:44)
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.model.RosterPosition.endXmlElement(RosterPosition.java:520)
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.xml.XmlHandler.endElement(XmlHandler.java:56)
2024-10-17T08:40:16.336 -------- ---- 	at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.endElement(AbstractSAXParser.java:618)
2024-10-17T08:40:16.336 -------- ---- 	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanEndElement(XMLDocumentFragmentScannerImpl.java:1728)
2024-10-17T08:40:16.336 -------- ---- 	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2899)
2024-10-17T08:40:16.336 -------- ---- 	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
2024-10-17T08:40:16.336 -------- ---- 	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)
2024-10-17T08:40:16.336 -------- ---- 	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)
2024-10-17T08:40:16.336 -------- ---- 	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)
2024-10-17T08:40:16.336 -------- ---- 	at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
2024-10-17T08:40:16.336 -------- ---- 	at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)
2024-10-17T08:40:16.336 -------- ---- 	at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:637)
2024-10-17T08:40:16.336 -------- ---- 	at com.fumbbl.ffb.xml.XmlHandler.parse(XmlHandler.java:90)
2024-10-17T08:40:16.336 -------- ---- 	... 17 more
```

Reply:
RosterPosition:520 is when the skill tag is evaluated. This code runs every time I load a game in standalone without any issue. That means the rules and factories are initialized. There seemsto be something else causing this error on your side.

![image](https://github.com/user-attachments/assets/f83de930-81ba-4fbe-bed7-d212d0ba263a)

Reply:
Hmm, weird. I wonder if it is somehow related to how I start the server, so something isn't initialized correctly. But it isn't showing up in the logs then. How do you start the server? 

I am using IntelliJ and is using this run configuration, which is building the jar and running it: 
![image](https://github.com/user-attachments/assets/11ba868e-aac3-4cae-8458-5ec3ecdabb6e)

Reply:
There is not much in the ini file to be adjusted apart from some file paths and connection properties. Can you double check if the issue still persists? And if it really only happens with the roster you downloaded from fumbbl?

Reply:
One difference is that I am running an orcale jdk 8. But that should not make much of a difference.

Reply:
I think I found the problem. If you start the client with a pre-defined teamId things do work correctly, while if you don't it crashes like I experienced. It is unrelated to the type of XML format. It also happens with the existing teams.

```
// Works
-player -coach BattleLore -teamId teamChaosBattleLore -teamName TestTeam

// Crash
-player -coach BattleLore
```

Looking at the code, that can be explained  as `joinWithTeam()` does initialize the rules: https://github.com/christerk/ffb/blob/d277aaf938985bb0118c8f707829bdeccab4acdd/ffb-server/src/main/java/com/fumbbl/ffb/server/handler/ServerCommandHandlerJoinApproved.java#L200

But it was missing from `sendTeamList()`.

Reply:
I see, then we should probably also call initRulesDependentMemebers to be consistent.

Reply:
Done. I think all comments should be addressed now.